### PR TITLE
feat: add scalar quantizer

### PR DESCRIPTION
**Pull request by BubbleCal**
plan to support scalar quantization
it's been without any optimization yet

### CODE REVIEW

File: rust/lance-index/src/vector/sq.rs [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+        sq.lower_bound = bounds.0;
+        sq.upper_bound = bounds.1;
+        sq
+    }
+
+    pub fn update_bounds(
+        &mut self,
+        vectors: &FixedSizeListArray,
+    ) -> Result<(T::Native, T::Native)> {
+        let data = vectors
```
Comment (eddyxu) [inline rust/lance-index/src/vector/sq.rs:91]: Does it mean that the training need scan over all data ?

File: rust/lance-index/src/vector/sq.rs [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+                ),
+                location: location!(),
+            })?
+            .as_slice();
+
+        // TODO: support SQ4
+        let builder: Vec<u8> = data
+            .iter()
+            .map(|v| {
+                let range = self.upper_bound - self.lower_bound;
```
Comment (eddyxu) [inline rust/lance-index/src/vector/sq.rs:148]: Should we have upper/lower bound for each dimension?

File: rust/lance-index/src/vector/sq.rs [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+        }
+    }
+
+    pub fn with_bounds(
+        num_bits: u16,
+        metric_type: MetricType,
+        bounds: (T::Native, T::Native),
+    ) -> Self {
+        let mut sq = Self::new(num_bits, metric_type);
+        sq.lower_bound = bounds.0;
```
Comment (eddyxu) [inline rust/lance-index/src/vector/sq.rs:82]: Isn't upper / lower bound per dimension?

File: rust/lance-index/src/vector/sq.rs [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+            metric_type,
+            lower_bound: T::MAX,
+            upper_bound: T::MIN,
+        }
+    }
+
+    pub fn with_bounds(
+        num_bits: u16,
+        metric_type: MetricType,
+        bounds: (T::Native, T::Native),
```
Comment (eddyxu) [inline rust/lance-index/src/vector/sq.rs:79]: Should we use range instead of tuple here

File: rust/lance-index/src/vector/sq.rs (hunk 2 above)
Comment (BubbleCal) [reply rust/lance-index/src/vector/sq.rs:148]: that will make the contribution of each dimension different

File: rust/lance-index/src/vector/sq.rs (hunk 1 above)
Comment (BubbleCal) [reply rust/lance-index/src/vector/sq.rs:84]: we don't have to, trained on samples to get the boundary, for the values out of range, we can simply map them to the lower/upper bound

**Review by eddyxu — APPROVED** [review]

**Pull request merged**